src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs

```diff
@@ -0,0 +1,87 @@
+// <copyright file="LazyInstrumentationLoader.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using OpenTelemetry.AutoInstrumentation.Configuration;
+using OpenTelemetry.AutoInstrumentation.Logging;
+
+namespace OpenTelemetry.AutoInstrumentation.Loading
+{
+    internal class LazyInstrumentationLoader : IDisposable
+    {
+        private static readonly ILogger Logger = OtelLogging.GetLogger();
+
+        private readonly List<AssemblyLoadDetector> _assemblyLoadDetectors = new();
+        private readonly ConcurrentBag<object> _instrumentations = new();
```

Comment:
maybe `ConcurrentBag<IDisposable>` or even `ConcurrentBag<Action>` ?

Reply:
We will need to track every instance, but not all of them are `IDisposable`.

Reply:
So why do we need them?

Reply:
Need to be sure that these instances are not garbage collected and live until SDK is disposed.

Reply:
please add a comment in code and resolve 😉